pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -767,3 +837,120 @@ func findBestHint(hints []topologymanager.TopologyHint) *topologymanager.Topolog
 	}
 	return &bestHint
 }
+
+func (p *staticPolicy) updateStateToReuse(s state.State, pod *v1.Pod, container *v1.Container, memoryBlocks []state.Block) {
+	podUID := string(pod.UID)
+
+	// If pod entries to p.reusableState other than the current pod exist, delete them
+	for uid := range p.reusableState {
+		if podUID != uid {
+			delete(p.reusableState, podUID)
+		}
+	}
+
+	if isInitContainer(pod, container) {
+		// update the reusable state to the state before the init containers allocation
+		if _, ok := p.reusableState[podUID]; !ok {
+			p.reusableState[podUID] = s.GetMachineState()
+		}
+		return
+	}
+
+	if _, ok := p.reusableState[podUID]; !ok {
+		return
+	}
+
+	// Otherwise it is an app container.
+	// Update the re-usable machine state
+	for _, memoryBlock := range memoryBlocks {
+		p.updateMachineState(p.reusableState[podUID], memoryBlock.NUMAAffinity, memoryBlock.Type, memoryBlock.Size)
+	}
+}
+
+func (p *staticPolicy) updateInitContainersReusableMemory(pod *v1.Pod, container *v1.Container, memoryBlocks []state.Block) {
+	podUID := string(pod.UID)
+
+	if isInitContainer(pod, container) {
+		if _, ok := p.initContainersReusableMemory[podUID]; !ok {
+			p.initContainersReusableMemory[podUID] = []map[string][]state.Block{}
+			p.initContainersReusableMemory[podUID] = append(p.initContainersReusableMemory[podUID], map[string][]state.Block{
+				container.Name: memoryBlocks,
+			})
+			return
+		}
+
+		p.initContainersReusableMemory[podUID] = append(p.initContainersReusableMemory[podUID], map[string][]state.Block{
+			container.Name: memoryBlocks,
+		})
+	}
+
+	// update old init containers memory, the idea that we will reduce memory requested by init containers from the old
+	// to the new one according to the current container memory request
+	for _, block := range memoryBlocks {
+		requestedSize := block.Size
+		for i := range p.initContainersReusableMemory[podUID] {
+			if requestedSize == 0 {
+				break
+			}
+
+			for containerName, containerBlocks := range p.initContainersReusableMemory[podUID][i] {
+				if requestedSize == 0 {
+					break
+				}
+
+				if containerName == container.Name {
+					continue
+				}
+
+				for j, containerBlock := range containerBlocks {
+					if block.Type != containerBlock.Type {
+						continue
+					}
+
+					if containerBlock.Size == 0 {
+						continue
+					}
+
+					if !isNUMAAffinitiesEqual(block.NUMAAffinity, containerBlock.NUMAAffinity) {
+						continue
+					}
+
+					if containerBlock.Size >= requestedSize {
+						containerBlocks[j].Size -= requestedSize
+						requestedSize = 0
+					} else {
+						requestedSize -= containerBlock.Size
+						containerBlocks[j].Size = 0
+					}
+				}
+			}
+		}
+	}
+}
+
+func isNUMAAffinitiesEqual(numaAffinity1, numaAffinity2 []int) bool {
```

Comment:
When / why are this affinities converted from a `bitmask.BitMask` type to an `[]int`? I ask because if they were still their original `bitmask` type, then you could just use `IsEqual()`:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/bitmask/bitmask.go#L135

Reply:
Maybe keep this function, but implement the body of it as a conversion to `bitmask` and a call to `isEqual()`.

Reply:
makes sense, will do it